File: Leetcode/21.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        ListNode *ret = nullptr, *p = l1, *q = l2;
        if (p == nullptr) {
            ret = q;
            if (q != nullptr) q = q->next;
        } else if (q == nullptr) {
            ret = p;
            p = p->next;
        } else {
            if (p->val < q->val) {
                ret = p;
                p = p->next;
            } else {
                ret = q;
                q = q->next;
            }
        }
        ListNode *k = ret;
        while (!(p == nullptr && q == nullptr)) {
            if (p == nullptr) {
                k->next = q;
                q = q->next;
            } else if (q == nullptr) {
                k->next = p;
                p = p->next;
            } else {
                if (p->val < q->val) {
                    k->next = p;
                    p = p->next;
                } else {
                    k->next = q;
                    q = q->next;
                }
            }
            k = k->next;
        }

        return ret;
    }
};

File: Leetcode/21.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        ret = ListNode(0)
        p = ret

        while not (l1 == None and l2 == None):
            if l1 == None:
                p.next = l2
                l2 = l2.next
            elif l2 == None:
                p.next = l1
                l1 = l1.next
            elif l1.val < l2.val:
                p.next = l1
                l1 = l1.next
            else:
                p.next = l2
                l2 = l2.next
            p = p.next

        return ret.next
